diff --git a/src/include/gdb_if.h b/src/include/gdb_if.h
index 332683c2..e594178c 100644
--- a/src/include/gdb_if.h
+++ b/src/include/gdb_if.h
@@ -21,10 +21,10 @@
 #ifndef INCLUDE_GDB_IF_H
 #define INCLUDE_GDB_IF_H
 
-#if PC_HOSTED == 0 && !defined(NO_LIBOPENCM3)
-#include <libopencm3/usb/usbd.h>
-void gdb_usb_out_cb(usbd_device *dev, uint8_t ep);
-#endif
+//#if PC_HOSTED == 0 && !defined(NO_LIBOPENCM3)
+//#include <libopencm3/usb/usbd.h>
+//void gdb_usb_out_cb(usbd_device *dev, uint8_t ep);
+//#endif
 
 int gdb_if_init(void);
 char gdb_if_getchar(void);
