diff --git a/src/gdb_main.c b/src/gdb_main.c
index 7901f21..28e85d2 100644
--- a/src/gdb_main.c
+++ b/src/gdb_main.c
@@ -58,13 +58,15 @@ typedef struct
 
 static char pbuf[BUF_SIZE+1];
 
-static target *cur_target;
+target *cur_target;
 static target *last_target;
 
 static void handle_q_packet(char *packet, int len);
 static void handle_v_packet(char *packet, int len);
 static void handle_z_packet(char *packet, int len);
 
+extern bool lnInterceptCommand( const char *packet); 
+
 static void gdb_target_destroy_callback(struct target_controller *tc, target *t)
 {
 	(void)tc;
@@ -110,6 +112,8 @@ int gdb_main_loop(struct target_controller *tc, bool in_syscall)
 		SET_IDLE_STATE(1);
 		size = gdb_getpacket(pbuf, BUF_SIZE);
 		SET_IDLE_STATE(0);
+
+		if(lnInterceptCommand(pbuf)) continue;
 		switch(pbuf[0]) {
 		/* Implementation of these is mandatory! */
 		case 'g': { /* 'g': Read general registers */
