diff --git a/src/target/target.c b/src/target/target.c
index c8a10a5..ac10c56 100644
--- a/src/target/target.c
+++ b/src/target/target.c
@@ -193,7 +193,7 @@ void target_add_flash(target *t, struct target_flash *f)
 	f->next = t->flash;
 	t->flash = f;
 }
-
+#if 0
 static ssize_t map_ram(char *buf, size_t len, struct target_ram *ram)
 {
 	return snprintf(buf, len, "<memory type=\"ram\" start=\"0x%08"PRIx32
@@ -229,7 +229,7 @@ bool target_mem_map(target *t, char *tmp, size_t len)
 		return false;
 	return true;
 }
-
+#endif
 static struct target_flash *flash_for_addr(target *t, uint32_t addr)
 {
 	for (struct target_flash *f = t->flash; f; f = f->next)
