diff --git a/src/target/cortexm.c b/src/target/cortexm.c
index 2288063..3ec76cc 100644
--- a/src/target/cortexm.c
+++ b/src/target/cortexm.c
@@ -63,7 +63,7 @@ const struct command_s cortexm_cmd_list[] = {
 	{"vector_catch", (cmd_handler)cortexm_vector_catch, "Catch exception vectors"},
 	{NULL, NULL, NULL}
 };
-
+extern const struct command_s *cortex_ln_commands; // MEANX
 /* target options recognised by the Cortex-M target */
 #define	TOPT_FLAVOUR_V6M	(1<<0)	/* if not set, target is assumed to be v7m */
 #define	TOPT_FLAVOUR_V7MF	(1<<1)	/* if set, floating-point enabled. */
@@ -355,6 +355,8 @@ bool cortexm_probe(ADIv5_AP_t *ap)
 	t->breakwatch_clear = cortexm_breakwatch_clear;
 
 	target_add_commands(t, cortexm_cmd_list, cortexm_driver_str);
+	target_add_commands(t, cortex_ln_commands, "extra freertos commands"); // MEANX
+
 
 	/* Probe for FP extension */
 	uint32_t cpacr = target_mem_read32(t, CORTEXM_CPACR);
