diff --git a/src/gdb_main.c b/src/gdb_main.c
index 7901f21..76f14fb 100644
--- a/src/gdb_main.c
+++ b/src/gdb_main.c
@@ -58,7 +58,7 @@ typedef struct
 
 static char pbuf[BUF_SIZE+1];
 
-static target *cur_target;
+target *cur_target;
 static target *last_target;
 
 static void handle_q_packet(char *packet, int len);
@@ -451,6 +451,13 @@ static void exec_q_crc(const char *packet, int len)
 			gdb_putpacket_f("C%lx", crc);
 	}
 }
+//----
+extern void execqOffsets(const char *packet, int len);
+extern void execqSymbol(const char *packet, int len);
+extern void execqThreadInfo(const char *packet, int len);
+extern void execqfThreadInfo(const char *packet, int len);
+extern void execqsThreadInfo(const char *packet, int len);
+
 static const cmd_executer q_commands[]=
 {
 	{"qRcmd,",                         exec_q_rcmd},
@@ -458,7 +465,14 @@ static const cmd_executer q_commands[]=
 	{"qXfer:memory-map:read::",        exec_q_memory_map},
 	{"qXfer:features:read:target.xml:",exec_q_feature_read},
 	{"qCRC:",                          exec_q_crc},
-
+//--
+	{"qOffsets"                       ,execqOffsets},
+	{"qSymbol:"                       ,execqSymbol},
+	{"qThreadInfo"                    ,execqThreadInfo},
+	{"qfThreadInfo"                   ,execqfThreadInfo},
+	{"qsThreadInfo"                   ,execqsThreadInfo},
+
+//--
 	{NULL,NULL},
 };
 
