diff --git a/src/gdb_main.c b/src/gdb_main.c
index bdff229..57603ec 100644
--- a/src/gdb_main.c
+++ b/src/gdb_main.c
@@ -56,7 +56,7 @@ static target *last_target;
 static void handle_q_packet(char *packet, int len);
 static void handle_v_packet(char *packet, int len);
 static void handle_z_packet(char *packet, int len);
-
+extern bool lnProcessCommand(int size, const char *data); // meanx
 static void gdb_target_destroy_callback(struct target_controller *tc, target *t)
 {
 	(void)tc;
@@ -322,7 +322,10 @@ int gdb_main_loop(struct target_controller *tc, bool in_syscall)
 			ERROR_IF_NO_TARGET();
 			handle_z_packet(pbuf, size);
 			break;
-
+        case 'H': 
+            if(lnProcessCommand(size,pbuf)) // MEANX
+                       continue;
+            // there is no break here, not a mistake
 		default: 	/* Packet not implemented */
 			DEBUG_GDB("*** Unsupported packet: %s\n", pbuf);
 			gdb_putpacketz("");
@@ -421,8 +424,11 @@ handle_q_packet(char *packet, int len)
 			gdb_putpacket_f("C%lx", crc);
 
 	} else {
+            if(!lnProcessCommand(len,packet))
+            {
 		DEBUG_GDB("*** Unsupported packet: %s\n", packet);
 		gdb_putpacket("", 0);
+            }
 	}
 }
 
