diff --git a/src/include/target.h b/src/include/target.h
index 5cf04ac..68467df 100644
--- a/src/include/target.h
+++ b/src/include/target.h
@@ -55,7 +55,9 @@ unsigned int target_designer(target *t);
 unsigned int target_idcode(target *t);
 
 /* Memory access functions */
+#if 0
 bool target_mem_map(target *t, char *buf, size_t len);
+#endif
 int target_mem_read(target *t, void *dest, target_addr src, size_t len);
 int target_mem_write(target *t, target_addr dest, const void *src, size_t len);
 /* Flash memory access functions */
@@ -179,4 +181,3 @@ struct target_controller {
 };
 
 #endif
-
