diff --git a/src/gdb_main.c b/src/gdb_main.c
index bdff229..f2f5a04 100644
--- a/src/gdb_main.c
+++ b/src/gdb_main.c
@@ -392,9 +392,16 @@ handle_q_packet(char *packet, int len)
 			gdb_putpacketz("E01");
 			return;
 		}
+
+#if 0 // MEANX, dont hammer the stack
 		char buf[1024];
 		target_mem_map(cur_target, buf, sizeof(buf)); /* Fixme: Check size!*/
 		handle_q_string_reply(buf, packet + 23);
+#else
+    char *out=ztarget_mem_map(cur_target);
+    handle_q_string_reply(out, packet + 23);
+    free(out);
+#endif
 
 	} else if (strncmp (packet, "qXfer:features:read:target.xml:", 31) == 0) {
 		/* Read target description */
