diff --git a/src/gdb_main.c b/src/gdb_main.c
index 9cca19e0..c192d474 100644
--- a/src/gdb_main.c
+++ b/src/gdb_main.c
@@ -459,6 +459,8 @@ static void exec_q_supported(const char *packet, const size_t length)
 		GDB_PACKET_BUFFER_SIZE);
 }
 
+extern  char *ztarget_mem_map(const target_s *t);
+
 static void exec_q_memory_map(const char *packet, const size_t length)
 {
 	(void)length;
@@ -471,9 +473,16 @@ static void exec_q_memory_map(const char *packet, const size_t length)
 		gdb_putpacketz("E01");
 		return;
 	}
+#if 0 // dont abuse the stack
 	char buf[1024];
 	target_mem_map(target, buf, sizeof(buf)); /* Fixme: Check size!*/
 	handle_q_string_reply(buf, packet);
+#else
+    char *out=ztarget_mem_map(cur_target);
+    handle_q_string_reply(out,packet);
+    free(out);
+	
+#endif
 }
 
 static void exec_q_feature_read(const char *packet, const size_t length)
