diff --git a/src/remote.c b/src/remote.c
index 77837b0a..9661114d 100644
--- a/src/remote.c
+++ b/src/remote.c
@@ -308,7 +308,7 @@ static void remote_packet_process_general(gdb_packet_s *const packet)
 		break;
 	case REMOTE_START:
 #if ENABLE_DEBUG == 1 && defined(PLATFORM_HAS_DEBUG)
-		debug_bmp = true;
+//		debug_bmp = true;
 #endif
 #ifndef PLATFORM_IDENT_DYNAMIC
 		remote_respond_string(REMOTE_RESP_OK, BOARD_IDENT);
