diff --git a/src/command.c b/src/command.c
index 52a60c6c..8bc57960 100644
--- a/src/command.c
+++ b/src/command.c
@@ -54,9 +54,9 @@
 static bool cmd_version(target_s *target, int argc, const char **argv);
 static bool cmd_help(target_s *target, int argc, const char **argv);
 
-static bool cmd_jtag_scan(target_s *target, int argc, const char **argv);
-static bool cmd_swd_scan(target_s *target, int argc, const char **argv);
-static bool cmd_auto_scan(target_s *target, int argc, const char **argv);
+/* static  */bool cmd_jtag_scan(target_s *target, int argc, const char **argv);
+/* static  */bool cmd_swd_scan(target_s *target, int argc, const char **argv);
+/* static  */bool cmd_auto_scan(target_s *target, int argc, const char **argv);
 static bool cmd_frequency(target_s *target, int argc, const char **argv);
 static bool cmd_targets(target_s *target, int argc, const char **argv);
 static bool cmd_morse(target_s *target, int argc, const char **argv);
@@ -230,7 +230,7 @@ bool cmd_help(target_s *target, int argc, const char **argv)
 	return true;
 }
 
-static bool cmd_jtag_scan(target_s *target, int argc, const char **argv)
+/*static*/ bool cmd_jtag_scan(target_s *target, int argc, const char **argv)
 {
 	(void)target;
 	(void)argc;
