diff --git a/src/target/adi.c b/src/target/adi.c
index 054ac516..469dec7a 100644
--- a/src/target/adi.c
+++ b/src/target/adi.c
@@ -874,22 +874,22 @@ void adi_ap_component_probe(
 			DEBUG_INFO("%s-> cortexm_probe\n", indent + 1);
 			cortexm_probe(ap);
 			break;
-		case aa_cortexa:
-			DEBUG_INFO("%s-> cortexa_probe\n", indent + 1);
-			cortexa_probe(ap, base_address);
-			break;
-		case aa_cortexr:
-			DEBUG_INFO("%s-> cortexr_probe\n", indent + 1);
-			cortexr_probe(ap, base_address);
-			break;
-		case aa_cortexa_armv8:
-			DEBUG_INFO("%s-> cortexa_armv8_dc_probe\n", indent + 1);
-			cortexa_armv8_dc_probe(ap, base_address);
-			break;
-		case aa_cti_armv8:
-			DEBUG_INFO("%s-> cortexa_armv8_cti_probe\n", indent + 1);
-			cortexa_armv8_cti_probe(ap, base_address);
-			break;
+			// MEANX 		case aa_cortexa:
+			// MEANX DEBUG_INFO("%s-> cortexa_probe\n", indent + 1);
+			// MEANX cortexa_probe(ap, base_address);
+			// MEANX break;
+			// MEANX case aa_cortexr:
+			// MEANX DEBUG_INFO("%s-> cortexr_probe\n", indent + 1);
+			// MEANX cortexr_probe(ap, base_address);
+			// MEANX break;
+			// MEANX case aa_cortexa_armv8:
+			// MEANX DEBUG_INFO("%s-> cortexa_armv8_dc_probe\n", indent + 1);
+			// MEANX cortexa_armv8_dc_probe(ap, base_address);
+			// MEANX break;
+			// MEANX case aa_cti_armv8:
+			// MEANX DEBUG_INFO("%s-> cortexa_armv8_cti_probe\n", indent + 1);
+			// MEANX cortexa_armv8_cti_probe(ap, base_address);
+			// MEANX break;
 		/* Handle when the component is a CoreSight component ROM table */
 		case aa_rom_table:
 			if (pidr & PIDR_SIZE_MASK)
