diff --git a/src/target/riscv32.c b/src/target/riscv32.c
index 251530d0..18b84694 100644
--- a/src/target/riscv32.c
+++ b/src/target/riscv32.c
@@ -99,7 +99,8 @@ bool riscv32_probe(target_s *const target)
 		break;
 	case JEP106_MANUFACTURER_WCH:
 		PROBE(ch32v003x_probe);
-		PROBE(ch32vx_probe);
+		//PROBE(ch32vx_probe);
+		PROBE(ch32v3xx_probe);
 		break;
 	}
 
diff --git a/src/target/target_probe.h b/src/target/target_probe.h
index 955ea5c2..67f721a2 100644
--- a/src/target/target_probe.h
+++ b/src/target/target_probe.h
@@ -115,5 +115,6 @@ bool stm32wb0_probe(target_s *target);
 bool zynq7_probe(target_s *target);
 
 void lpc55_dp_prepare(adiv5_debug_port_s *dp);
+bool ch32v3xx_probe(target_s *target);
 
 #endif /* TARGET_TARGET_PROBE_H */
