diff --git a/src/target/riscv32.c b/src/target/riscv32.c
index b993805d..343c24b5 100644
--- a/src/target/riscv32.c
+++ b/src/target/riscv32.c
@@ -30,7 +30,6 @@
  * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
  * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
  */
-
 #include "general.h"
 #include "target.h"
 #include "target_internal.h"
@@ -92,6 +91,13 @@ static void riscv32_regs_write(target_s *target, const void *data);
 static int riscv32_breakwatch_set(target_s *target, breakwatch_s *breakwatch);
 static int riscv32_breakwatch_clear(target_s *target, breakwatch_s *breakwatch);
 
+static const char *riscv32_target_arch(target_s *target)
+{
+	(void)target;
+	return "RISCV";
+}
+
+#define PROBED_ARCH TOPT_ARCH_RISCV
 bool riscv32_probe(target_s *const target)
 {
 	/* 'E' base ISA has 16 GPRs + PC, 'I' base ISA has 32 GPRs + PC */
@@ -827,7 +833,7 @@ static int riscv32_breakwatch_set(target_s *const target, breakwatch_s *const br
 	}
 	/* Grab the address to set the breakwatch on and configure the hardware */
 	const uint32_t address = breakwatch->addr;
-	const bool result = riscv_config_trigger(hart, trigger, mode, &config, &address);
+	const bool result = riscv_config_trigger32(hart, trigger, mode, config, address);
 	/* If configuration succeeds, store the trigger index in the breakwatch structure */
 	if (result)
 		breakwatch->reserved[0] = trigger;
@@ -840,5 +846,82 @@ static int riscv32_breakwatch_clear(target_s *const target, breakwatch_s *const
 	riscv_hart_s *const hart = riscv_hart_struct(target);
 	const uint32_t config = RV32_MATCH_ADDR_DATA_TRIGGER;
 	const uint32_t address = 0;
-	return riscv_config_trigger(hart, breakwatch->reserved[0], RISCV_TRIGGER_MODE_UNUSED, &config, &address) ? 0 : -1;
+	return riscv_config_trigger32(hart, breakwatch->reserved[0], RISCV_TRIGGER_MODE_UNUSED, config, address) ? 0 : -1;
+}
+
+/*
+ *   Small helper function to translate target to hart and simplify parameters
+ *   We assume it is 32 bits
+ */
+static inline bool riscv32_target_csr_write(target_s *target, const uint16_t reg, uint32_t val)
+{
+	riscv_hart_s *const hart = riscv_hart_struct(target);
+	return riscv_csr_write(hart, reg, &val);
+}
+
+/*
+ *   Small helper function to translate target to hart and simplify parameters
+ *   We assume it is 32 bits
+ */
+static inline bool riscv32_target_csr_read(target_s *target, const uint16_t reg, uint32_t *val)
+{
+	riscv_hart_s *const hart = riscv_hart_struct(target);
+	return riscv_csr_read(hart, reg, val);
+}
+
+/*
+ *   Execute code on the target with the signature void function(a,b,c,d)
+ *       - codexec is the address the code to tun is located at
+ *       - param1/2/3/4 will end up as the 4 parameters of the stub function
+ *
+ *   The flashstub must not use the stack at all.
+ *   It returns true on success, false on error
+ *   There is a built-in timeout of 10 seconds
+ */
+bool riscv32_run_stub(
+	target_s *target, uint32_t loadaddr, uint32_t param1, uint32_t param2, uint32_t param3, uint32_t param4)
+{
+	bool ret = false;
+	uint32_t pc = 0;
+	uint32_t mie = 0;
+	uint32_t zero = 0;
+	// save PC & MIE
+	target->reg_read(target, RISCV_REG_PC, &pc, 4);
+	riscv32_target_csr_read(target, RV_CSR_MIE, &mie);
+	riscv32_target_csr_write(target, RV_CSR_MIE, zero); // disable interrupt
+	target->reg_write(target, RISCV_REG_A0, &param1, 4);
+	target->reg_write(target, RISCV_REG_A1, &param2, 4);
+	target->reg_write(target, RISCV_REG_A2, &param3, 4);
+	target->reg_write(target, RISCV_REG_A3, &param4, 4);
+	target->reg_write(target, RISCV_REG_PC, &loadaddr, 4);
+
+	target->halt_resume(target, false); // go!
+	platform_timeout_s timeout;
+	platform_timeout_set(&timeout, 10000);
+	target_halt_reason_e reason = TARGET_HALT_RUNNING;
+	while (reason == TARGET_HALT_RUNNING) {
+		if (platform_timeout_is_expired(&timeout)) {
+			reason = TARGET_HALT_ERROR;
+			break;
+		}
+		reason = target->halt_poll(target, NULL);
+	}
+	switch (reason) {
+	case TARGET_HALT_REQUEST:
+		ret = true;
+		break;
+	case TARGET_HALT_ERROR: // room left here for more detailed handling
+	default:
+		break;
+	}
+	target->halt_request(target);
+	if (ret) {
+		uint32_t a0;
+		target->reg_read(target, RISCV_REG_A0, &a0, 4);
+		ret = (a0 == 0);
+	}
+	// restore PC & MIE
+	riscv32_target_csr_write(target, RV_CSR_MIE, mie); // put back MIE
+	target->reg_write(target, RISCV_REG_PC, &pc, 4);
+	return ret;
 }
diff --git a/src/target/riscv_debug.h b/src/target/riscv_debug.h
index 85b2ce52..f25c51e5 100644
--- a/src/target/riscv_debug.h
+++ b/src/target/riscv_debug.h
@@ -306,6 +306,14 @@ typedef struct riscv_hart {
 #define RV_FPU_GDB_OFFSET     33
 #define RV_FPU_GDB_CSR_OFFSET 66
 
+// General purpose registers
+#define RISCV_REG_A0 10
+#define RISCV_REG_A1 11
+#define RISCV_REG_A2 12
+#define RISCV_REG_A3 13
+#define RISCV_REG_PC 32
+#define RISCV_REG_SP 2
+
 /* JTAG DTM function declarations */
 #ifdef CONFIG_RISCV
 void riscv_jtag_dtm_handler(uint8_t dev_index);
@@ -330,6 +338,8 @@ bool riscv_csr_write(riscv_hart_s *hart, uint16_t reg, const void *data);
 riscv_match_size_e riscv_breakwatch_match_size(size_t size);
 bool riscv_config_trigger(
 	riscv_hart_s *hart, uint32_t trigger, riscv_trigger_state_e mode, const void *config, const void *address);
+bool riscv_config_trigger32(
+	riscv_hart_s *hart, uint32_t trigger, riscv_trigger_state_e mode, uint32_t config, uint32_t address);
 
 bool riscv_attach(target_s *target);
 void riscv_detach(target_s *target);
@@ -341,4 +351,15 @@ uint32_t riscv32_pack_data(const void *src, uint8_t access_width);
 void riscv32_mem_read(target_s *target, void *dest, target_addr64_t src, size_t len);
 void riscv32_mem_write(target_s *target, target_addr64_t dest, const void *src, size_t len);
 
+/*
+ *   Execute code on the target with the signature void stub_function(a,b,c,d)
+ *       - codexec is the address the code to run is located at
+ *       - param1/2/3/4 will end up as the 4 parameters of the stub function
+ *       
+ *   The flashstub must not use the stack at all.
+ *   The flashstub must return 0 on success, not 0 on error
+ *   There is a built-in timeout of 10 seconds
+ */
+bool riscv32_run_stub(target_s *target, uint32_t loadaddr, uint32_t a0, uint32_t a1, uint32_t a2, uint32_t a3);
+
 #endif /*TARGET_RISCV_DEBUG_H*/
