diff --git a/src/target/riscv_debug.c b/src/target/riscv_debug.c
index 0047b91f..11d07bce 100644
--- a/src/target/riscv_debug.c
+++ b/src/target/riscv_debug.c
@@ -947,6 +947,37 @@ uint8_t riscv_mem_access_width(const riscv_hart_s *const hart, const target_addr
 	return access_width;
 }
 
+/* All possible match modes */
+#define RV32_MATCH_MASK 0x00000007U
+
+bool riscv_config_trigger32(riscv_hart_s *const hart, const uint32_t trigger, const riscv_trigger_state_e mode,
+	const uint32_t config, const uint32_t address)
+{
+	/*
+	 * Select the trigger and write the new configuration to it provided by config.
+	 * tdata1 (RV_TRIG_DATA_1) becomes mcontrol (match control) for this -
+	 * see §5.2.9 pg53 of the RISC-V debug spec v0.13.2 for more details.
+	 */
+	bool result = riscv_csr_write(hart, RV_TRIG_SELECT | RV_CSR_FORCE_32_BIT, &trigger) &&
+		riscv_csr_write(hart, RV_TRIG_DATA_1, &config) && riscv_csr_write(hart, RV_TRIG_DATA_2, &address);
+	if (result) {
+		/* If that succeeds, verify the mode was taken into account */
+		uint32_t read_back;
+		result = riscv_csr_read(hart, RV_TRIG_DATA_1, &read_back) &&
+			((read_back & RV32_MATCH_MASK) == (config & RV32_MATCH_MASK));
+		if (!result) {
+			read_back = 0;
+			riscv_csr_write(hart, RV_TRIG_DATA_1, &read_back);
+		}
+	}
+	if (result) {
+		/* If that succeeds, update the slot with the new mode it's in */
+		hart->trigger_uses[trigger] &= ~RV_TRIGGER_MODE_MASK;
+		hart->trigger_uses[trigger] |= mode;
+	}
+	return result;
+}
+
 static void riscv_hart_discover_triggers(riscv_hart_s *const hart)
 {
 	/* Discover how many breakpoints this hart supports */
