diff --git a/src/include/gdb_packet.h b/src/include/gdb_packet.h
index 72f25898..2a58a08a 100644
--- a/src/include/gdb_packet.h
+++ b/src/include/gdb_packet.h
@@ -46,9 +46,9 @@
 #define GDB_PACKET_RETRIES 3U /* Number of times to retry sending a packet */
 
 #if defined(__MINGW32__) || defined(__MINGW64__) || defined(__CYGWIN__)
-#define GDB_FORMAT_ATTR __attribute__((format(__MINGW_PRINTF_FORMAT, 1, 2)))
+#define GDB_FORMAT_ATTR 
 #elif defined(__GNUC__) || defined(__clang__)
-#define GDB_FORMAT_ATTR __attribute__((format(printf, 1, 2)))
+#define GDB_FORMAT_ATTR 
 #else
 #define GDB_FORMAT_ATTR
 #endif
