diff --git a/src/target/target.c b/src/target/target.c
index 315f9837..e19dffb7 100644
--- a/src/target/target.c
+++ b/src/target/target.c
@@ -697,3 +697,30 @@ void tc_printf(target_s *target, const char *fmt, ...)
 	// MEANX (void)fflush(stdout);
 	va_end(ap);
 }
+
+// MEANX
+
+bool breakpoint_available(target_s *target)
+{
+	return !target->no_hw_breakpoint;
+}
+
+void target_breakpoint_watchpoint_count(target_s *const target, uint32_t *brk, uint32_t *wtch)
+{
+	if (target->breakpoint_watchpoint_count) {
+		target->breakpoint_watchpoint_count(target, brk, wtch);
+		return;
+	}
+	*brk = 0;
+	*wtch = 0;
+}
+
+bool target_has_hw_breakpoint(target_s *const target)
+{
+	return !target->no_hw_breakpoint;
+}
+
+const sw_breakpoint_helpers *target_get_mw_helpers(target_s *const target)
+{
+	return target->sw_breakpoint_helpers;
+}
