diff --git a/src/target/target_internal.h b/src/target/target_internal.h
index 33d23deb..d792799a 100644
--- a/src/target/target_internal.h
+++ b/src/target/target_internal.h
@@ -119,6 +119,19 @@ struct breakwatch {
 
 #define MAX_CMDLINE 81
 
+/* 
+ * These option functions are meant to be used by software breakpoint.
+ * In particular they should aim at minimal write size (page) and
+ * they must not disturb too much the target. NO FLASHSTUB 
+ */
+struct sw_breakpoint_helpers_s {
+	uint32_t (*page_size)(target_s *t);
+	bool (*page_erase)(target_s *target, uint32_t addr);
+	bool (*page_write)(target_s *target, uint32_t addr, const uint8_t *src, uint32_t page_size);
+};
+
+typedef struct sw_breakpoint_helpers_s sw_breakpoint_helpers;
+
 typedef void (*priv_free_func)(void *flash);
 
 struct target {
@@ -160,6 +173,14 @@ struct target {
 	bool (*enter_flash_mode)(target_s *target);
 	bool (*exit_flash_mode)(target_s *target);
 
+	/* CRC32 function */
+	bool (*crc32)(target_s *target, target_addr32_t start_adress, size_t size, uint32_t *crc32);
+
+	/* Software breakpoint helpers */
+	const struct sw_breakpoint_helpers_s *sw_breakpoint_helpers;
+	bool no_hw_breakpoint;
+	void (*breakpoint_watchpoint_count)(target_s *const target, uint32_t *brk, uint32_t *wtch);
+
 	/* Target-defined options */
 	uint32_t target_options;
 
@@ -205,11 +226,15 @@ struct target {
 extern const command_s platform_cmd_list[];
 #endif
 
+const sw_breakpoint_helpers *target_get_mw_helpers(target_s *const target);
+void target_breakpoint_watchpoint_count(target_s *const target, uint32_t *brk, uint32_t *wtch);
+bool target_has_hw_breakpoint(target_s *const target);
 void target_print_progress(platform_timeout_s *timeout);
 void target_ram_map_free(target_s *target);
 void target_flash_map_free(target_s *target);
 void target_mem_map_free(target_s *target);
 void target_add_commands(target_s *target, const command_s *cmds, const char *name);
+bool breakpoint_available(target_s *target);
 void target_add_ram32(target_s *target, target_addr32_t start, uint32_t len);
 void target_add_ram64(target_s *target, target_addr64_t start, uint64_t len);
 void target_add_flash(target_s *target, target_flash_s *flash);
@@ -237,7 +262,7 @@ bool target_check_error(target_s *target);
 #if defined(__MINGW32__) || defined(__MINGW64__) || defined(__CYGWIN__)
 #define TC_FORMAT_ATTR __attribute__((format(__MINGW_PRINTF_FORMAT, 2, 3)))
 #elif defined(__GNUC__) || defined(__clang__)
-#define TC_FORMAT_ATTR __attribute__((format(printf, 2, 3)))
+#define TC_FORMAT_ATTR //MEANX__attribute__((format(printf, 2, 3)))
 #else
 #define TC_FORMAT_ATTR
 #endif
