diff --git a/src/platforms/hosted/platform.h b/src/platforms/hosted/platform.h
index e220166d..aae6dd59 100644
--- a/src/platforms/hosted/platform.h
+++ b/src/platforms/hosted/platform.h
@@ -48,7 +48,9 @@ void platform_buffer_flush(void);
 	do {                 \
 	} while (0)
 #define PLATFORM_HAS_POWER_SWITCH
+#ifndef PLATFORM_HAS_RVSWD
 #define PLATFORM_HAS_RVSWD
+#endif
 
 #define PRODUCT_ID_ANY 0xffffU
 
