diff --git a/src/target/cortex.h b/src/target/cortex.h
index 8f116683..d92be2f1 100644
--- a/src/target/cortex.h
+++ b/src/target/cortex.h
@@ -77,13 +77,22 @@
 #define CORTEX_CPUID_REVISION_MASK 0x00f00000U
 #define CORTEX_CPUID_PATCH_MASK    0x0000000fU
 
-#define CORTEX_FLOAT_REG_COUNT          33U
-#define CORTEX_DOUBLE_REG_COUNT         17U
-#define CORTEXM_GENERAL_REG_COUNT       20U /* General purpose register count for Cortex-M cores */
-#define CORTEXM_TRUSTZONE_REG_COUNT     4U  /* TrustZone register count for Cortex-M cores */
-#define CORTEXAR_GENERAL_REG_COUNT      17U /* General purpose register count for Cortex-A/R cores */
+#define CORTEX_FLOAT_REG_COUNT  33U
+#define CORTEX_DOUBLE_REG_COUNT 17U
+// meanx #define CORTEXM_GENERAL_REG_COUNT       20U /* General purpose register count for Cortex-M cores */
+// meanx #define CORTEXM_TRUSTZONE_REG_COUNT     4U  /* TrustZone register count for Cortex-M cores */
+// meanx #define CORTEXAR_GENERAL_REG_COUNT      17U /* General purpose register count for Cortex-A/R cores */
 #define CORTEXA_ARMV8_GENERAL_REG_COUNT 31U /* General purpose register count for Cortex-A/R cores in AArch64 mode */
 
+// meanx
+#define CORTEXM_GP_REG_COUNT        13U /* General purpose register count for Cortex-M cores r0 to r12 */
+#define CORTEXM_SP_REG_COUNT        04U /* sp lr pc xpsr */
+#define CORTEXM_SYS_REG_COUNT       03U /* Sys section : MSP/PSP/state */
+#define CORTEXM_TRUSTZONE_REG_COUNT 10U /* TrustZone register count for Cortex-M cores msp psp msplim psplim state x2 */
+// obsolete, for compatibility with older lpc code
+#define CORTEXM_GENERAL_REG_COUNT 20U
+// / meanx
+
 adiv5_access_port_s *cortex_ap(target_s *target);
 
 #endif /* TARGET_CORTEX_H */