diff --git a/src/include/target.h b/src/include/target.h
index ce0be5d1..31c63602 100644
--- a/src/include/target.h
+++ b/src/include/target.h
@@ -38,6 +38,14 @@ typedef SSIZE_T ssize_t;
 typedef int32_t mode_t;
 #endif /* _MSC_VER */
 
+// MEANX
+#define TOP_ARCH_SHIFT   28
+#define TOPT_GET_ARCH(x) ((x >> TOP_ARCH_SHIFT) & 3)
+#define TOPT_SET_ARCH(val, arch) 		val = (val & ~(3 << TOP_ARCH_SHIFT)) | (arch << TOP_ARCH_SHIFT);
+#define TOPT_ARCH_ARM   1
+#define TOPT_ARCH_RISCV 2
+// /MEANX
+
 typedef struct target target_s;
 typedef uint32_t target_addr32_t;
 typedef uint64_t target_addr64_t;
diff --git a/src/target/target_probe.h b/src/target/target_probe.h
index 4463906d..e76c9036 100644
--- a/src/target/target_probe.h
+++ b/src/target/target_probe.h
@@ -23,14 +23,19 @@
 #include "target.h"
 #include "adiv5.h"
 
+/* -- MEANX -- */
+void bmp_tweak_options(uint32_t *opt);
+
 #define STRINGIFY(x) #x
 /* Probe launch macro used by the CPU-generic layers to then call CPU-specific routines safely */
-#define PROBE(x)                                    \
-	do {                                            \
-		DEBUG_TARGET("Calling %s\n", STRINGIFY(x)); \
-		if ((x)(target))                            \
-			return true;                            \
-		target_check_error(target);                 \
+#define PROBE(x)                                          \
+	do {                                                  \
+		DEBUG_TARGET("Calling %s\n", STRINGIFY(x));       \
+		if ((x)(target)) {                                \
+			bmp_tweak_options(&(target->target_options)); \
+			return true;                                  \
+		}                                                 \
+		target_check_error(target);                       \
 	} while (0)
 
 /*
@@ -113,5 +118,6 @@ bool stm32c5_probe(target_s *target);
 bool zynq7_probe(target_s *target);
 
 void lpc55_dp_prepare(adiv5_debug_port_s *dp);
+bool ch32v3xx_probe(target_s *target);
 
 #endif /* TARGET_TARGET_PROBE_H */

