diff --git a/src/remote.c b/src/remote.c
index 215611d0..b48cf124 100644
--- a/src/remote.c
+++ b/src/remote.c
@@ -318,7 +318,7 @@ static void remote_packet_process_general(char *packet, const size_t packet_len)
 		break;
 	case REMOTE_START:
 #if ENABLE_DEBUG == 1 && defined(PLATFORM_HAS_DEBUG)
-		debug_bmp = true;
+//		debug_bmp = true;
 #endif
 		remote_respond_string(REMOTE_RESP_OK, PLATFORM_IDENT "" FIRMWARE_VERSION);
 		break;
