diff --git a/src/target/riscv_debug.h b/src/target/riscv_debug.h
index 8af394a7..a4eaff15 100644
--- a/src/target/riscv_debug.h
+++ b/src/target/riscv_debug.h
@@ -234,4 +234,13 @@ uint32_t riscv32_pack_data(const void *src, uint8_t access_width);
 void riscv32_mem_read(target_s *target, void *dest, target_addr_t src, size_t len);
 void riscv32_mem_write(target_s *target, target_addr_t dest, const void *src, size_t len);
 
+#define RISCV_REG_A0 10
+#define RISCV_REG_A1 11
+#define RISCV_REG_A2 12
+#define RISCV_REG_PC 32
+#define RISCV_REG_SP 2
+
+
+bool riscv32_run_stub(target_s *target, uint32_t loadaddr, uint32_t a0, uint32_t a1, uint32_t a2, uint32_t temporary_stack);
+
 #endif /*TARGET_RISCV_DEBUG_H*/
