diff --git a/src/remote.c b/src/remote.c
index c7591b51..0458693b 100644
--- a/src/remote.c
+++ b/src/remote.c
@@ -315,7 +315,7 @@ static void remote_packet_process_general(unsigned i, char *packet)
 #endif
 	case REMOTE_START:
 #if defined(ENABLE_DEBUG) && defined(PLATFORM_HAS_DEBUG)
-		debug_bmp = true;
+		//debug_bmp = true;
 #endif
 		remote_respond_string(REMOTE_RESP_OK, PLATFORM_IDENT "" FIRMWARE_VERSION);
 		break;
