diff --git a/src/target/target_internal.h b/src/target/target_internal.h
index 313a4fdb..a396c1f7 100644
--- a/src/target/target_internal.h
+++ b/src/target/target_internal.h
@@ -211,7 +211,7 @@ bool target_check_error(target_s *target);
 #if defined(__MINGW32__) || defined(__MINGW64__) || defined(__CYGWIN__)
 #define TC_FORMAT_ATTR __attribute__((format(__MINGW_PRINTF_FORMAT, 2, 3)))
 #elif defined(__GNUC__) || defined(__clang__)
-#define TC_FORMAT_ATTR __attribute__((format(printf, 2, 3)))
+#define TC_FORMAT_ATTR //MEANX__attribute__((format(printf, 2, 3)))
 #else
 #define TC_FORMAT_ATTR
 #endif
